### PR TITLE
Corrected decoding function

### PR DESCRIPTION
**Pull request by tomkocmi**
Za zhruba 4 dny bych sem hodil porovnani s puvodnim kodem.

### CODE REVIEW

**Comment by tomkocmi** [comment]
Vypada to, ze to zadny rozdil neudelalo, mam az to dobehne porovnat i cas nebo tohle vylepseni melo pouze zlepsit skore? A pokud to neprida zadne skore, tak ma se to mergovat?

**Comment by jlibovicky** [comment]
Má se to mergovat, protože předtím to špatně (resp. normalizované přibližně), tohle je přesně.

**Comment by tomkocmi** [comment]
Myslim, ze takto opraveny script neni kompatibilni s Beam searchem:
 Inputs to operation while/attention_decoder/attention_encoder/Attention/Select of type Select must have the same size and shape.  Input 0: [1,17] != input 1: [20,17]

Jak to opravit?

**Comment by jlibovicky** [comment]
To je pro mě vždycky záhada, kdy se ta batch implicitně broadcastuje a kdy to nefunguje. Každopádně se to vyřeší jako je to v attention combination: https://github.com/ufal/neuralmonkey/blob/master/neuralmonkey/encoders/encoder_wrapper.py#L291
je potřeba zjistit batch size udělat to explicitně pomocí `tf.tile`.

File: neuralmonkey/decoding_function.py [hunk 1]
```diff
@@ -178,9 +179,10 @@ def attention(self, decoder_state: tf.Tensor,
             if self.input_weights is None:
                 weights = tf.nn.softmax(s)
             else:
-                weights_all = tf.nn.softmax(s) * self.input_weights
-                norm = tf.reduce_sum(weights_all, 1, keep_dims=True) + 1e-8
-                weights = weights_all / norm
+                masked_logits = tf.where(
+                    tf.equal(self.input_weights, 1),
+                    s, -np.inf * tf.ones_like(s))
```
Comment (jindrahelcl) [inline neuralmonkey/decoding_function.py]: zkus sem dát místo toho ones_like tohle:

```
tf.fill(dims=tf.shape(s), value=-np.inf)
```

a mohlo by to i broadcastovat

**Comment by tomkocmi** [comment]
@jindrahelcl zpusob nefungoval, chvili jsem si hral s tim tile, ale to se mi take nedarilo. Mrknu na to v tydnu, musim presne pochopit kde je ten problem s batchi. Zkousel jsem asi uplnou blbost:

tiled = tf.tile(s, [tf.shape(s)[0], 1])
masked_logits = tf.where(tf.equal(self.input_weights, 1), tiled, -np.inf * tf.ones_like(tiled))

File: neuralmonkey/decoding_function.py (hunk 1 above)
Comment (jindrahelcl) [reply neuralmonkey/decoding_function.py]: proč to nefunguje? co to píše?

Comment (tomkocmi) [reply neuralmonkey/decoding_function.py]: porad stejnou hlasku ... mam totiz pocit ze si to nestezuje na ty np.inf ale na to kdyz to vraci pouze s, coz je ale divne

Comment (jindrahelcl) [reply neuralmonkey/decoding_function.py]: Stěžuje si operace `Select`, což je jméno pro to volání toho `where`, viz TF kód. Stěžuje si, že `s` a `-np.inf * tf.ones_like(s)` mají různý shape. `ones_like(s)`by mělo vracet shape jako má `s`. Takže to obal printama a zjisti, jaký shape mají tenzory `s`, `xxx = tf.ones_like(s)` a `-np.inf * xxx` a snad ti to dá nějakej insight.  Rovnou zjisti i jakej shape má ten `tf.equal(...)` - ten where se chová jinak, když má condition jednu dimenzi a když jich má víc.

Bude tam demence v tom, že v prvnim kroku beam search je batch size jedna (vycházim ze stejnýho stavu a se stejným vstupem a vemu beam nejlepších) a v ostatních krocích je batch size velikost beamu.

**Comment by jlibovicky** [comment]
Mělo by to být tak, že těch logitů už je stejně jako batche/beamu v dekodéru, protože tam vstupuje nabatchovaný stav dekodéru 
a ten shape se správně broadcastuje (nejspíš tady: https://github.com/ufal/neuralmonkey/blob/master/neuralmonkey/decoding_function.py#L204). V prvním argumentu `tf.where` ale vybíráš podle masky, která má první dimenzi 1, protože enkodér má v ruce jenom jeden example, takže to, co je potřeba "tilovat" je tam maska. ne logity.

**Comment by tomkocmi** [comment]
aha, ono to chtelo maskovat vahy a ne masky (podle toho jak je to pojmenovane), to jsem spatne pochopil. Tak ted uz to bezi v pohode.

**Comment by jindrahelcl** [comment]
changes incorporated in #509, closing...
